doc/chips/9.rst

```diff
@@ -45,36 +49,285 @@ The idea of ``mason`` is based heavily off of Rust's cargo_. It is meant to
-(or at least becomes much more complicated if it is not). Nix also only works
-on linux/osx right now. Cygwin support is not ruled out, but not actively
-worked on.
 
-Example
--------
+Mason Pipeline
+--------------
 
-Lets go through an example from our standard modules and see how it would look
-as a package.
+The package manager pipeline has 4 key states, where each state acts as the
+input to the next state:
+
+* **Project Code**: ``src/``
+* **Manifest File**: ``Mason.toml``
+
+  * Markup-style file containing package metadata
+
+* **Lock File**: ``Mason.lock``
+
+  * File containing list of dependencies generated from ``Mason.toml``.
+  * Contains the necessary information to guarantee a consistent build on
+    all supported platforms.
+
+* **Dependency Code**: ``~/.mason/src/``
+
+  * Dependencies that are downloaded into a local directory.
+
+A package repository will consist of the following directory hierarchy:
+
+::
+
+  MyPackage/
+    Mason.toml
+    src/
+      main.chpl
+
+Upon invoking ``mason build``, the dependencies will be parsed into a directed
+acyclic graph, resolved, and serialized into ``Mason.lock``. They will then be
+downloaded into ``~/.mason/src/`` mirroring Cargo's ``~/.cargo/src/`` common pool
+of dependencies. Finally, if the package is an application, the binary will
+build into ``target/``. The final directory hierarchy will look like this:
+
+::
+
+  ~/.mason/src/
+    (dependencies of MyPackage)
+
+
+  MyPackage/
+    Mason.toml
+    Mason.lock
+    src/
+      main.chpl
+    target/
+      ( MyPackage )
+
+
+Project Code
+++++++++++++
+
+This is the Chapel source code written by the package author.
+
+For example, ``main.chpl``:
+
+.. code-block:: chapel
+
+   proc main() {
+       writeln("Hello, world!");
+    }
+
+Project Code -> Manifest File
++++++++++++++++++++++++++++++
+
+The user will be responsible for writing the manifest file in its entirety.
+
+Manifest File
++++++++++++++
+
+The ``Mason.toml`` manifest file will closely follow the design of
+`Cargo's manifest file <http://doc.crates.io/manifest.html>`_,
+``Cargo.toml``. Some fields will vary, if they are too Rust-specific or could
+benefit from being more Chapel-specific.
+
+For example, ``Mason.toml``:
+
+.. code-block:: ini
+
+   [package]
+   name = "hello_world"
+   version = "0.1.0"
+   authors = ["Bradford Chamberlain <brad@chamberlain.com>"]
+
+   [dependencies]
+   Curl = "1.0.0"
+
+**Registry**
+
+Eventually, we will want a service similar to Cargo's crates.io_ to serve as our
+registry. Due to the amount of effort behind such a task, it has been decided
+that this can come at a later stage.
+
+The initial mason registry would be a GitHub repository containing a list of
+versioned manifest files from packages, not too unlike that of the OS X
+Homebrew package manager registry.
+
+The registry would follow a hierarchy as follows:
+
+::
+
+  https://github.com/chapel-lang/mason-registry
```

Comment:
I think the mason tool will be part of the chapel-lang/chapel repository to begin with, but the registry should definitely be a separate repository. Otherwise, it would require all package contributing users to sign and turn in a CLA in order to check in their package.